app/src/main/java/org/stepik/android/domain/course_search/interactor/CourseSearchInteractor.kt

```diff
@@ -36,56 +51,72 @@ constructor(
             }
 
-    private fun fetchCourseSearchResultsDetails(searchResults: PagedList<SearchResult>): Observable<PagedList<CourseSearchResult>> {
+    fun getDiscussionThreads(step: Step): Single<List<DiscussionThread>> =
+        discussionThreadRepository
+            .getDiscussionThreads(*step.discussionThreads?.toTypedArray() ?: arrayOf())
+
+    private fun fetchCourseSearchResultsDetails(searchResults: PagedList<SearchResult>): Observable<PagedList<CourseSearchResultListItem.Data>> {
         val lessonIds = searchResults
             .mapNotNull { it.lesson }
-            .mapToLongArray { it }
-
-        val lessonOwners = searchResults
-            .mapNotNull { it.lessonOwner }
-            .toSet()
 
         val commentOwners = searchResults
             .mapNotNull { it.commentUser }
-            .toSet()
 
-        val combined = (lessonOwners + commentOwners).toList()
+        val stepIds = searchResults
+            .mapNotNull { it.step }
+            .mapToLongArray { it }
 
-        return lessonRepository
-            .getLessons(*lessonIds, primarySourceType = DataSourceType.CACHE)
-            .flatMapObservable { lessons ->
-                fetchProgressesAndUsers(searchResults, lessons, combined)
-            }
+        return Observables.zip(
+            lessonRepository.getLessons(*lessonIds.mapToLongArray { it }, primarySourceType = DataSourceType.REMOTE).toObservable(),
+            stepRepository.getSteps(*stepIds, primarySourceType = DataSourceType.REMOTE).toObservable()
+        ).flatMap { (lessons, steps) ->
+            val unitIds = lessons.flatMap { lesson -> lesson.unit.map { it } }
+            unitRepository
+                .getUnits(unitIds, DataSourceType.REMOTE)
+                .flatMapObservable { units ->
+                    fetchProgressesAndUsers(searchResults, lessons, units, steps, commentOwners)
+                }
+        }
     }
 
-    private fun fetchProgressesAndUsers(searchResults: PagedList<SearchResult>, lessons: List<Lesson>, userIds: List<Long>): Observable<PagedList<CourseSearchResult>> =
-        Observable.zip(
+    private fun fetchProgressesAndUsers(searchResults: PagedList<SearchResult>, lessons: List<Lesson>, units: List<Unit>, steps: List<Step>, userIds: List<Long>): Observable<PagedList<CourseSearchResultListItem.Data>> =
+        Observables.zip(
             progressRepository.getProgresses(lessons.getProgresses(), primarySourceType = DataSourceType.REMOTE).toObservable(),
-            userRepository.getUsers(userIds, primarySourceType = DataSourceType.REMOTE).toObservable()
-        ) { progresses, users ->
+            userRepository.getUsers(userIds, primarySourceType = DataSourceType.REMOTE).toObservable(),
+            sectionRepository.getSections(units.map(Unit::section), primarySourceType = DataSourceType.REMOTE).toObservable()
+        ).map { (progresses, users, sections) ->
```

Comment:
We can write `) { progresses, users, sections ->` in more efficient way without `map`